lib/Jsrt/Jsrt.cpp

```diff
@@ -2746,3 +3230,504 @@ CHAKRA_API JsRunSerializedScriptWithCallback(_In_ JsSerializedScriptLoadSourceCa
+            threadContext->TTDLog->PushMode(TTD::TTDMode::ExcludedExecution);
+            context->GetScriptContext()->InitializeCoreImage_TTD();
+            threadContext->TTDLog->PopMode(TTD::TTDMode::ExcludedExecution);
+
+            context->GetScriptContext()->InitializeDebuggingActionsAsNeeded_TTD();
+
+            scriptContext->InitializeDebugging();
+            context->GetScriptContext()->GetDebugContext()->GetProbeContainer()->InitializeInlineBreakEngine(runtime->GetJsrtDebugManager());
+            context->GetScriptContext()->GetDebugContext()->GetProbeContainer()->InitializeDebuggerScriptOptionCallback(runtime->GetJsrtDebugManager());
+            threadContext->GetDebugManager()->SetLocalsDisplayFlags(Js::DebugManager::LocalsDisplayFlags::LocalsDisplayFlags_NoGroupMethods);
+        }
+        catch(...)
+        {
+            //something went horribly wrong
+
+            *targetStartSnapTime = -1;
+        }
+    }
+
+    return (*targetStartSnapTime != -1) ? JsNoError : JsErrorFatal;
+#endif
+}
+
+CHAKRA_API JsTTDMoveToTopLevelEvent(_In_ INT64 snapshotTime, _In_ INT64 eventTime)
+{
+#if !ENABLE_TTD_DEBUGGING
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+    if(scriptContext->GetThreadContext()->TTDLog == nullptr)
+    {
+        AssertMsg(false, "Should only happen in TT debugging mode.");
+        return JsErrorCategoryUsage;
+    }
+
+    TTD::EventLog* elog = scriptContext->GetThreadContext()->TTDLog;
+    JsErrorCode res = JsNoError;
+
+    //a special indicator to use the time from the argument flag
+    if(eventTime == -2)
+    {
+        eventTime = scriptContext->GetThreadContext()->TTDLog->GetKthEventTime(Js::Configuration::Global.flags.TTDStartEvent);
+    }
+
+    try
+    {
+        elog->PushMode(TTD::TTDMode::ExcludedExecution);
+        BEGIN_JS_RUNTIME_CALLROOT_EX(scriptContext, false)
+        {
+            elog->DoSnapshotInflate(snapshotTime);
+        }
+        END_JS_RUNTIME_CALL(scriptContext);
+        elog->PopMode(TTD::TTDMode::ExcludedExecution);
+
+        //If the log has a BP requested then we should set the actual bp here
+        if(scriptContext->GetThreadContext()->TTDLog->HasPendingTTDBP())
+        {
+            TTD::TTDebuggerSourceLocation bpLocation;
+            scriptContext->GetThreadContext()->TTDLog->GetPendingTTDBPInfo(bpLocation);
+
+            Js::FunctionBody* body = bpLocation.ResolveAssociatedSourceInfo(scriptContext);
+            Js::Utf8SourceInfo* utf8SourceInfo = body->GetUtf8SourceInfo();
+
+            charcount_t charPosition;
+            charcount_t byteOffset;
+            utf8SourceInfo->GetCharPositionForLineInfo((charcount_t)bpLocation.GetLine(), &charPosition, &byteOffset);
+            long ibos = charPosition + bpLocation.GetColumn() + 1;
```

Comment:
Yes, only way to have a pending BP should be via TTLog reverse step and this should only be possible from the debugger.